src/Swift.Runtime/src/Library/Arc.cs

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Runtime.InteropServices;
+
+namespace Swift.Runtime;
+
+/// <summary>
+/// Arc is a class containing p/invokes for Swift Automatic Reference Counting and memory management.
+/// </summary>
+public static class Arc {
+    const string swiftCore = "/usr/lib/swift/libswiftCore.dylib";
+
+    /// <summary>
+    /// Retain a heap-allocated Swift object
+    /// </summary>
+    /// <param name="p">Pointer to an unmanaged Swift object, must be non-null.</param>
+    [DllImport(swiftCore, CallingConvention = CallingConvention.Cdecl)]
+    static extern void swift_retain(IntPtr p);
+
+    /// <summary>
+    /// Retain a heap-allocated Swift object
+    /// </summary>
+    /// <param name="p">Pointer to an unmanaged Swift object, must be non-null.</param>
+    /// <returns>The pointer passed in.</returns>
+    /// <exception cref="ArgumentNullException">Throws if p is null</exception>
+    public static IntPtr Retain(IntPtr p)
+    {
+        if (p == IntPtr.Zero)
+            throw new ArgumentNullException(nameof(p));
```

Comment:
```suggestion
        ArgumentNullException.ThrowIfNull(p);
```
(dtto for other places)

Reply:
This is illegal without an unsafe cast to `void *`.

Reply:
The right overload is internal unfortunately: https://github.com/dotnet/runtime/blob/d6be63071b06354d581060e633cdb64da307d37c/src/libraries/System.Private.CoreLib/src/System/ArgumentNullException.cs#L87